domain/cloud/state/state.go

```diff
@@ -332,62 +334,65 @@ ON CONFLICT(region_uuid, key) DO UPDATE
 	q := `
-SELECT cloud.uuid, cloud.name, cloud_type_id, 
-       cloud.endpoint, cloud.identity_endpoint, 
-       cloud.storage_endpoint, skip_tls_verify, 
-       auth_type.type, cloud_type.type,
-       COALESCE((SELECT true
+WITH controllers AS (SELECT model_metadata.cloud_uuid
         FROM model_metadata
 		INNER JOIN user
         ON user.uuid = model_metadata.owner_uuid
-        WHERE cloud_uuid = cloud.uuid
-        AND model_metadata.name = ?
-        AND user.name = ?), false) AS controller_cloud
+        AND model_metadata.name = $M.controller_name
+        AND user.name = $M.controller_user_name)
+SELECT (cloud.uuid, cloud.name, cloud_type_id, 
+       cloud.endpoint, cloud.identity_endpoint, 
+       cloud.storage_endpoint, skip_tls_verify) AS (&Cloud.*),
+       IIF(controllers.cloud_uuid IS NULL, false, true) AS &Cloud.is_controller_cloud,
+       auth_type.type AS &M.cloud_auth_type, cloud_type.type AS &M.cloud_type
 FROM   cloud
        LEFT JOIN cloud_auth_type 
             ON cloud.uuid = cloud_auth_type.cloud_uuid
        JOIN auth_type 
             ON auth_type.id = cloud_auth_type.auth_type_id
        JOIN cloud_type 
             ON cloud_type.id = cloud.cloud_type_id
+       LEFT JOIN controllers
+            ON controllers.cloud_uuid = cloud.uuid
 `
 
-	args := []any{
-		coremodel.ControllerModelName,
-		coremodel.ControllerModelOwnerUsername,
-	}
 	if name != "" {
-		q = q + "WHERE cloud.name = ?"
-		args = append(args, name)
+		q = q + "WHERE cloud.name = $M.cloud_name"
+	}
+
+	loadCloudStmt, err := sqlair.Prepare(q, sqlair.M{}, Cloud{})
+	if err != nil {
+		return nil, errors.Trace(err)
 	}
 
-	rows, err := tx.QueryContext(ctx, q, args...)
-	if err != nil && err != sql.ErrNoRows {
+	// The cloud_name entry in the map will be ignored if the WHERE clause above
+	// is not in the query.
+	args := sqlair.M{
+		"controller_name":      coremodel.ControllerModelName,
+		"controller_user_name": coremodel.ControllerModelOwnerUsername,
+		"cloud_name":           name,
+	}
+	iter := tx.Query(ctx, loadCloudStmt, args).Iter()
+	if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
 		return nil, errors.Trace(err)
 	}
-	defer func() { _ = rows.Close() }()
+	defer func() { _ = iter.Close() }()
 
 	clouds := make(map[string]*cloud.Cloud)
-	for rows.Next() {
-		var (
-			dbCloud       Cloud
-			cloudType     string
-			cloudAuthType string
-		)
-		if err := rows.Scan(
-			&dbCloud.ID, &dbCloud.Name, &dbCloud.TypeID, &dbCloud.Endpoint, &dbCloud.IdentityEndpoint,
-			&dbCloud.StorageEndpoint, &dbCloud.SkipTLSVerify, &cloudAuthType, &cloudType, &dbCloud.IsControllerCloud,
-		); err != nil {
+	m := sqlair.M{}
+	for iter.Next() {
+		var dbCloud Cloud
+		if err := iter.Get(&dbCloud, m); err != nil {
 			return nil, errors.Trace(err)
 		}
 		cld, ok := clouds[dbCloud.ID]
 		if !ok {
 			cld = &cloud.Cloud{
 				Name:              dbCloud.Name,
-				Type:              cloudType,
+				Type:              m["cloud_type"].(string),
```

Comment:
This can potentially panic. Check it exists before using it.

Reply:
Added an `ok` check. I was unsure about this when writing it, technically SQLair is guaranteeing that this value is there, it would have returned an error when the query was executed if it was not, but I agree we should avoid any potential for panics at all here.